src/components/Purchase/CheckoutComplete.js

```diff
@@ -0,0 +1,120 @@
+import React, { useState, useEffect } from 'react';
+import { Container, Paper, Typography, Grid, Box } from '@mui/material';
+import style from './CheckoutComplete.module.css';
+import { useSearchParams } from 'react-router-dom';
+import Loader from '../layout/Loader/Loader';
+import axiosInstance from '../../apis/axiosClient';
+import CheckoutMessageCard from './CheckoutMessageCard';
+
+const ordersEndpoint = '/orders';
+
+const CheckoutComplete = () => {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const clientSecret = searchParams.get('payment_intent_client_secret');
+  const orderId = searchParams.get('orderId');
+
+  const [isLoading, setIsLoading] = useState(true);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [orderData, setOrderData] = useState('');
+  const [isPaymentSuccessful, setIsPaymentSuccessful] = useState(false);
+
+  // Callback function to update isPaymentSuccessful
+  const updatePaymentStatus = (isSuccess) => {
+    setIsPaymentSuccessful(isSuccess);
+  };
+
+  useEffect(() => {
+    const orderDetails = async (id) => {
+      setIsLoading(true);
+      setErrorMessage('');
+      try {
+        const response = await axiosInstance.get(`${ordersEndpoint}/${id}`);
+        setIsLoading(false);
+        setOrderData(response.data?.order);
+        console.log(response.data?.order);
+      } catch (error) {
+        setIsLoading(false);
+        setErrorMessage(
+          `An error occurred during loading order details with id ${id}. ${error.message}. ${error.request.responseText}.`
+        );
+      }
+    };
+    orderDetails(orderId);
+  }, [clientSecret]);
+
+  // Calculate the total number of items purchased
+  let totalItemsPurchased = 0;
+  if (orderData && orderData.orderItems) {
+    orderData.orderItems.forEach((item) => {
+      totalItemsPurchased += item.quantity;
+    });
+  }
+
+  return (
+    <>
+      <Container className={style.mainContainer}>
+        <Grid container spacing={3}>
+          <Grid item xs={12} sm={6}>
+            <CheckoutMessageCard
+              isLoading={isLoading}
+              clientSecret={clientSecret}
+              totalItemsPurchased={totalItemsPurchased}
+              updateStatus={updatePaymentStatus}
+            />
+          </Grid>
+
+          <Grid item xs={12} sm={6}>
+            <Paper className={`${style.paper} ${style.rightPaper}`}>
+              {!isPaymentSuccessful && (
+                <Typography variant="h6">
+                  Payment was unsuccessful! Sorry for the inconvenience.
+                </Typography>
```

Comment:
Right now; I get a flash of the "payment was unsuccessful" message before the real result loads because when the component first loads and we haven't yet gotten the info from stripe, `isPaymentSuccessful` starts off as false. Maybe it would make sense to only show this if we've finished loading:

```suggestion
              {!loading && !isPaymentSuccessful && (
                <Typography variant="h6">
                  Payment was unsuccessful! Sorry for the inconvenience.
                </Typography>
```

Reply:
 f

Reply:
Yes. Good point! The glitch was indeed present.

Reply:
Sorry for the continued back and forth on this but I think I'm still seeing the glitch :thinking: 

It seems to be because we disable the `isLoading` state as soon as the Order API response has been returned even if the payment status from stripe hasn't been returned yet. One solution might be to have two separate loading states:

```javascript
// CheckoutComplete.js

  const [isLoading, setIsLoading] = useState(true);
  const [isPaymentLoading, setIsPaymentLoading] = useState(true);
  // this variable "derives state" from the above to state variables. 
  // It will evaluate to true when both the order and the  payment status have been loaded
  const bothLoaded = !isLoading && !isPaymentLoading;
  
  // Callback function to update isPaymentSuccessful
  const updatePaymentStatus = (isSuccess) => {
    setIsPaymentLoading(false);
    setIsPaymentSuccessful(isSuccess);
  };
  
  // in the return block show the loader if both responses haven't loaded yet, otherwise check for payment status or errors:
            <Grid item xs={12} sm={6}>
            <Paper className={`${style.paper} ${style.rightPaper}`}>
              {!bothLoaded && <Loader className="small-spinner" />}
              {bothLoaded && (
                <>
                ...etc
                </>
```

```javascript
// PaymentStatus.js

// Before, this was only ever either true or false. 
// But that meant that our parent component could not differentiate between "paymentsuccess is false because the payment failed" and "payment status is false because  we haven't loaded it yet and that's the default state"
// So we're essentially adding a third possible value for this state "null" to mean "payment status has not yet been retrieved"
// Then the parent components can rely on this to determine whether or not to set paymentSucces values yet
  const [isSuccess, setIsSuccess] = useState(null); 
```

```javascript
// CheckoutMessageCard.js

// Same idea as in PaymentStatus, start off the value as "null" to indicate "not yet defined/loaded"

const [isPaymentSuccessful, setIsPaymentSuccessful] = useState(null);

// only update the payment successful state here and in the parent component if we actually have gotten a non-null value from PaymentStatus component
  
  // Callback function to update isPaymentSuccessful
  const updatePaymentStatus = (isSuccess) => {
    if (isSuccess !== null) {
      setIsPaymentSuccessful(isSuccess);
    }
  };

  useEffect(() => {
    if (isPaymentSuccessful !== null) {
      updateStatus(isPaymentSuccessful);
    }
  }, [updateStatus, isPaymentSuccessful]);
```

Reply:
Hi Akos, I think I found another alternative way to fix the loading glitch. I decided to take the initiative and offer my own solution. It seems to me that it has a little less lines of code, however, I can implement your approach if my solution does not look optimal enough. In my approach, instead of creating another loading state, I simply pass the value from isLoading through a prop to the child component. Yes, this is prop drilling, but on the other hand, it seems to me that in this way we add less lines of code. Your opinion on this matter is very interesting to me.

Reply:
I reverted my changes and implemented the previously recommended approach.